### DISCUSSION TITLE
How to make make a container with a wrapping layout?

### DISCUSSION THREAD

**Discussion by Cationiz3r** [hecrj/iced#2013] [Q&A] [answered]
Using the `Responsive` widget, I was able to make a grid like layout from a vector of fixed-size images. Since their width are known before rendering, the number of images per row can be calculated.

However, say, I want to use buttons with text instead, since the texts can have arbitrary width, is there a way I can retrieve the length of each button? Maybe another approach is necessary?

Here's a demo in HTML of what I am trying to recreate in iced.

![image](https://github.com/iced-rs/iced/assets/51510836/f8881e94-343b-4189-91bc-a572b7a5ad44)

```html
<!DOCTYPE html>
<html lang="en">
	<head>
		<title>Wrapping demo</title>
		<meta charset="UTF-8">
		<meta name="viewport" content="width=device-width, initial-scale=1">
		<link href="css/style.css" rel="stylesheet">
	</head>
	<body>
		<div id="container">
			<div class="tag">Lorem</div>
			<div class="tag">ipsum</div>
			<div class="tag">dolor</div>
			<div class="tag">sit</div>
			<div class="tag">amet</div>
			<div class="tag">consectetur</div>
			<div class="tag">adipisicing</div>
			<div class="tag">elit.</div>
			<div class="tag">Nam</div>
			<div class="tag">eligendi</div>
			<div class="tag">beatae</div>
			<div class="tag">veniam!</div>
			<div class="tag">Nesciunt</div>
			<div class="tag">inventore</div>
			<div class="tag">veniam</div>
			<div class="tag">magnam</div>
			<div class="tag">sed</div>
			<div class="tag">illum</div>
			<div class="tag">molestiae</div>
			<div class="tag">nisi,</div>
			<div class="tag">amet</div>
			<div class="tag">dolor.</div>
			<div class="tag">Cumque</div>
			<div class="tag">alias,</div>
			<div class="tag">suscipit</div>
			<div class="tag">iste</div>
			<div class="tag">a</div>
			<div class="tag">illum</div>
			<div class="tag">ad</div>
			<div class="tag">quisquam?</div>
		</div>
		<style>
			html, body {
				margin: 0;
				padding: 0;
				height: 100%;
			}
			body {
				display: flex;
				align-items: center;
				justify-content: center;
			}
			#container {
				display: flex;
				flex-wrap: wrap;
				width: 50%;
			}
			.tag {
				margin: 0 5px 5px 0;
				padding: 5px;
				background-color: #ccc;
				font-weight: bold;
				border-radius: 3px;
			}
		</style>
	</body>
</html>
```

**Comment by veniamin-ilmer** [comment]
Inside of the text class, I see this method which seems to calculate the length of the text inside:

    fn layout(
        &self,
        renderer: &Renderer,
        limits: &layout::Limits,
    ) -> layout::Node {
        let limits = limits.width(self.width).height(self.height);

        let size = self.size.unwrap_or_else(|| renderer.default_size());

        let bounds = renderer.measure(
            &self.content,
            size,
            self.line_height,
            self.font.unwrap_or_else(|| renderer.default_font()),
            limits.max(),
            self.shaping,
        );

        let size = limits.resolve(bounds);

        layout::Node::new(size)
    }

**Comment by genusistimelord** [comment, marked as answer]
you can also use https://docs.rs/iced_aw/latest/iced_aw/native/wrap/index.html widget from iced_aw
